### PR TITLE
Fixed #31415 -- Fixed crash when nested OuterRef is used with operators or in database functions.

### PR DESCRIPTION
**Pull request by hramezani**
Fixes [ticket 31415](https://code.djangoproject.com/ticket/31415).

### CODE REVIEW

File: tests/expressions/tests.py [hunk 1]
```diff
@@ -576,6 +576,14 @@ def test_nested_subquery_outer_ref_with_autofield(self):
         # This exercises the double OuterRef form with AutoField as pk.
         self.assertCountEqual(outer, [first, second])
 
+    def test_nested_subquery_outer_ref_with_operator(self):
+        first = Time.objects.create(time='09:00')
+        second = Time.objects.create(time='17:00')
+        inner = Time.objects.filter(time__lt=OuterRef(OuterRef('time')) * 2).values('time')
+        middle = SimulationRun.objects.annotate(other=Subquery(inner)).values('other')[:1]
+        outer = Time.objects.annotate(other=Subquery(middle, output_field=IntegerField()))
+        self.assertCountEqual(outer, [first, second])
```
Comment (felixxm) [inline tests/expressions/tests.py:585]: Can we also add a test for reported use case?

File: tests/expressions/tests.py [hunk 2]
```diff
@@ -576,6 +576,33 @@ def test_nested_subquery_outer_ref_with_autofield(self):
...
+
+    def test_nested_subquery_outer_ref_with_function(self):
+        for item in Employee.objects.all():
+            print(item.firstname)
```
Comment (felixxm) [inline tests/expressions/tests.py:589]: :thinking:

Comment (hramezani) [reply tests/expressions/tests.py]: Sorry, those were for debugging.
I Removed them.

File: tests/expressions/tests.py (hunk 1 above)
Comment (charettes) [reply tests/expressions/tests.py:585]: I think this test can be entirely removed if `test_nested_subquery_outer_ref_with_function` covers the usage case appropriately.

Comment (felixxm) [reply tests/expressions/tests.py]: Agreed, removed.

**Comment by felixxm** [comment]
@hramezani Thanks :+1:

**Pull request merged**